Add support for the arm and wrist

jordan-powers opened the pull request:
- Support added for raising and lowering the arm, extending and retracting the wrist, and doing both while maintaining the wrist angle (needs tuning)
- CAN IDs are not final and some preferences need to be calibrated
- Controls are open to change and adjustment

dslit3 committed:
Add support for the arm and wrist

- Support added for raising and lowering the arm, extending and retracting the wrist, and doing both while maintaining the wrist angle (needs tuning)
- CAN IDs are not final and some preferences need to be calibrated
- Controls are open to change and adjustment

jordan-powers commented on src/main/java/frc/robot/subsystems/ArmSubsystem.java:
```diff
@@ -0,0 +1,61 @@
+
+    @Override
+    public void periodic() {
+        if (MotorMovement.isOutOfBounds(leftShoulder, SHOULDER_LIMIT)) {
+            leftShoulder.stopMotor();
+        }
+        if (MotorMovement.isOutOfBounds(rightShoulder, SHOULDER_LIMIT)) {
+            rightShoulder.stopMotor();
+        }
+        if (MotorMovement.isOutOfBounds(wrist, WRIST_LIMIT)) {
+            wrist.stopMotor();
+        }
```
I have two concerns here:

1. I don't like that there could be potentially a command telling the motor to go, while at the same time the subsystem periodic() is telling the motor to stop. Which the motor decides to do would depend on the execution order of commands vs. subsystem periodic. While we could in theory look at the scheduler documentation/source code, it feels like a footgun to rely on the implementation details of the scheduler. IMO, there should only ever be one place in the code that is setting the motors: maybe add a private `setMotor()` that does bounds checking, then call that method from `adjustArm()` and `adjustWrist()`.
In fact, I see that the `MotorMovement` helper already does this, which makes me question why you are doing it here as well?

3. I doubt `stopMotor()` will be enough to actually stop the arm/wrist from moving, since gravity will be acting to pull the mechanisms down. What we really need is velocity/position PID and trapezoidal motion planning. However, `stopMotor()` is probably good enough until we get that PID working.

jordan-powers commented on src/main/java/frc/robot/utils/MoPrefs.java:
```diff
@@ -17,6 +17,9 @@ public final class MoPrefs {
     public static Pref<Double> driveRampTime = doublePref("Drive Ramp Time", 0.25);
+    public static Pref<Double> shoulderSetpointRpm = doublePref("Shoulder Speed RPM", 200.0); // TODO: Find actual value
+    public static Pref<Double> wristSetpointRpm = doublePref("Wrist Speed RPM", 200.0); // TODO: Find actual value
+    public static Pref<Double> armJointSetpointRatio = doublePref("Wrist/Shoulder Speed Ratio", 1); // TODO: Find actual value
```
What is the `armJointSetpointRatio`? I don't see it used anywhere?

jordan-powers replied:
It is supposed to be used to keep the wrist parallel (maintain its angle relative to the ground) when raising: the ratio between the speed of the arm to the speed of the wrist

jordan-powers commented on src/main/java/frc/robot/utils/MotorMovement.java:
```diff
@@ -0,0 +1,29 @@
+            (setpoint < 0 && controller.getReverseLimitSwitch(limitType).isPressed());
+    }
+
+    public static boolean isOutOfBounds(CANSparkMax controller, Type limitType) {
+        return
+            controller.getForwardLimitSwitch(limitType).isPressed() ||
+            controller.getReverseLimitSwitch(limitType).isPressed();
+    }
```
We don't have limit switches this year. Instead, we're using the REV absolute encoders and the NEO built-in encoders (we need both since the absolute encoders have very coarse accuracy, whereas the built-in hall effect sensors are very accurate). You'll need to add some method of fusing the two sensors together, using the absolute encoder for zeroing and the built-in encoder for actual fine measurement.

jordan-powers commented on src/main/java/frc/robot/utils/MotorMovement.java:
```diff
@@ -0,0 +1,29 @@
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.SparkMaxLimitSwitch.Type;
+
+public final class MotorMovement {
```
I'm not convinced this needs to be its own class. Seems like it would be simple enough to just implement as a private `setMotor()` method on the ArmSubsystem.

jordan-powers commented on src/main/java/frc/robot/input/SingleControllerInput.java:
```diff
@@ -38,7 +48,31 @@ public double getTurnRequest() {
+        return controller.getAButton();
     }
 
+    @Override
+    public JoystickButton getRaiseArmButton() {
+        return this.raiseArm;
+    }
+
+    @Override
+    public JoystickButton getLowerArmButton() {
+        return this.lowerArm;
+    }
+
+    @Override
+    public JoystickButton getRetractWristButton() {
+        return this.retractWrist;
+    }
+
+    @Override
+    public JoystickButton getExtendWristButton() {
+        return this.extendWrist;
+    }
```
What's the rationale behind making these boolean inputs instead of analog inputs? We still have a couple of analog inputs left on the controller: the right y-axis and the shoulders. Wouldn't the increased precision of analog input help the driver more accurately position the shoulder/wrist?

jordan-powers commented on src/main/java/frc/robot/commands/RaiseArmCommand.java:
```diff
@@ -0,0 +1,48 @@
+    public RaiseArmCommand(ArmSubsystem arms, double power, BooleanSupplier parallel) {
+        this.arms = arms;
+        this.power = power;
+        this.shouldBeParallel = parallel;
```
What is `shouldBeParallel`?

jordan-powers commented on src/main/java/frc/robot/commands/RaiseArmCommand.java:
```diff
@@ -0,0 +1,48 @@
+        this.arms = arms;
+        this.power = power;
+        this.shouldBeParallel = parallel;
+    }
```
The `RaiseArmCommand` isn't requiring the ArmSubsystem? Commands, especially commands that move motors, should always require the subystem they're using, to prevent conflicting motor movements.

The same comment applies to the `RotateWristCommand`.

jordan-powers commented on src/main/java/frc/robot/commands/RotateWristCommand.java:
```diff
@@ -0,0 +1,29 @@
+    private final ArmSubsystem arms;
+    private final double power;
+
+    public RotateWristCommand(ArmSubsystem arms, double power) {
```
What's the rationale behind having one single `ArmSubsystem` for both the shoulder and wrist, but two separate commands `RotateArmCommand` and `RotateWristCommand`? We should decide if the arm and wrist are separate entities, or two parts of the same entity, then keep that decision consistent throughout the code.

jordan-powers commented on src/main/java/frc/robot/commands/RaiseArmCommand.java:
```diff
@@ -0,0 +1,48 @@
+
+    @Override
+    public void end(boolean interrupted) {
+        this.arms.idleArm();
+
+        if (this.wasParallel) {
+            this.arms.idleWrist();
+        }
+    }
```
A better pattern might be to have a third `IdleArm` command that tries to keep the arm from moving, and prevents it from extending beyond the frame perimeter. Set the `IdleArm` command as the default command of the `ArmSubsystem`, then have a `ArmTeleop` command that requires the `ArmSubsytem`. Then, when the `ArmTeleop` command finishes, the `IdleArm` is automatically scheduled, ensuring that there is always some code running to make sure the arm doesn't extend past the frame perimeter.

jordan-powers reviewed (commented):
A good start, but I have questions and see several problems.

jordan-powers committed:
Merge branch 'main' into arm-control

momentumrobotics-4999 committed:
Incomplete rework of the arm subsystem

momentumrobotics-4999 committed:
Finalize shift to spark max smart motion

jordan-powers commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -39,27 +42,33 @@ public class RobotContainer {
 
         drive.setDefaultCommand(driveCommand);
         visionSubsystem.setDefaultCommand(defaultVisionCommand);
+        arms.setDefaultCommand(Optional.ofNullable(armChooser.getSelected()).orElse(directArmControlCommand));
```
I like the idea of using a chooser for the arm mode!

However, I'm not sure about the timing here. If the default command is set right when the robot starts, there's no time for the chooser to be shown on the shuffleboard or the user to select a different value.

I'd recommend adding a listener to the chooser's backing Topic (or polling the selected option in `robotPeriodic()`) so that the defaultCommand is updated every time the chosen arm command changes.

jordan-powers commented on src/main/java/frc/robot/commands/TeleopArmCommand.java:
```diff
@@ -0,0 +1,45 @@
+        this.arms.adjustWrist(this.getWristPower(this.input));
+    }
+
+    public abstract double getShoulderPower(MoInput input);
+
+    public abstract double getWristPower(MoInput input);
```
Not sure what the use is of this abstract interface, since we only use explicit `shoulderPower` and `wristPower` values when operating in "direct" mode.

When we are using PID, the rio should not specify the power value for the shoulder or wrist motors. Setting power values is done by the PIDController built in to the SparkMax. All the rio does is call `SparkMaxPIDController.setReference( {position}, ControlType.SmartMotion)`

Essentially the command has two different outputs depending on what mode we're in:
- In "direct" mode: `shoulderPower` and `wristPower`
- In PID mode: `shoulderPosition` and `wristPosition`

Since these two outputs are so different, I don't see the utility of associating the two modes with an inheritance hierarchy.

dslit3 replied:
"power" doesn't seem like the correct word to describe it, but "setpoint" doesn't either
In velocity PID mode (Direct), those will return a value from 0-1 which will then be converted to a velocity setpoint (0 to Max RPM)
In position PID mode (Button Combos), those 0-1 values will be converted to an explicit position setpoint (all the way down to all the way up)

jordan-powers commented on src/main/java/frc/robot/input/SingleControllerInput.java:
```diff
@@ -38,7 +38,16 @@ public double getTurnRequest() {
 
+    @Override
+    public double getDirectShoulderRequest() {
+        return this.controller.getRightTriggerAxis() - this.controller.getLeftTriggerAxis();
```
Needs `deadzone` and `curve`.

The same applies to `getDirectWristRequest`

jordan-powers commented on src/main/java/frc/robot/subsystems/ArmSubsystem.java:
```diff
@@ -0,0 +1,49 @@
+    private final CANSparkMax wrist = new CANSparkMax(
+        Constants.ARM_WRIST.address, CANSparkMaxLowLevel.MotorType.kBrushless);
+
+    private final MoSparkMaxPID leftShoulderVelocityPID = new MoSparkMaxPID(Type.VELOCITY, leftShoulder);
+    private final MoSparkMaxPID rightShoulderVelocityPID = new MoSparkMaxPID(Type.VELOCITY, rightShoulder);
+    private final MoSparkMaxPID wristVelocityPID = new MoSparkMaxPID(Type.VELOCITY, wrist);
```
SmartMotion uses position pid, not velocity.

jordan-powers replied:
Actually, I see now that you're using the velocity PID for direct mode, and haven't implemented SmartMotion yet.

That's good: I agree we should focus on getting the simpler "direct" mode working before we worry about smartMotion.
However, as I mention in my other comment, we shouldn't be using PID in the fallback--instead just directly set the motor power.

jordan-powers commented on src/main/java/frc/robot/subsystems/ArmSubsystem.java:
```diff
@@ -0,0 +1,49 @@
+        Constants.ARM_WRIST.address, CANSparkMaxLowLevel.MotorType.kBrushless);
+
+    private final MoSparkMaxPID leftShoulderVelocityPID = new MoSparkMaxPID(Type.VELOCITY, leftShoulder);
+    private final MoSparkMaxPID rightShoulderVelocityPID = new MoSparkMaxPID(Type.VELOCITY, rightShoulder);
```
The left and right should should always move at exactly the same speed, since they are physically connected.
Thus we don't want two separate PID controllers running, since there's the possibility they might conflict and fight each other.

Only instantiate a MoSparkMaxPID for the left shoulder motor, and make the right shoulder motor follow it using  [CanSparkMax.follow()](https://codedocs.revrobotics.com/java/com/revrobotics/cansparkmax#follow(com.revrobotics.CANSparkMax,boolean)) (make sure to invert it if necessary)

jordan-powers commented on src/main/java/frc/robot/utils/MoSparkMaxPID.java:
```diff
@@ -0,0 +1,48 @@
+import com.revrobotics.SparkMaxPIDController;
+import com.revrobotics.CANSparkMax.ControlType;
+
+public class MoSparkMaxPID {
```
Overall, this looks good!

One suggestion for improvement: have MoSparkMaxPID implement the `GraphValues` interface. Then you don't need as many lambdas in `TunerUtils.forSparkMaxSmartMotion()`.

You could replace:
```java
.buildGraphValues()
.withSetpointFrom(sparkMax::getReference)
.withLastMeasurementFrom(sparkMax::getMeasurement)
.withLastOutputFrom(sparkMax::getOutput)
.finishGraphValues()
```
with
```java
.withGraphValues(sparkMax)
```

which, IMO, is a bit cleaner.

jordan-powers reviewed (commented):
Looking good! Just a couple more things I noticed

jordan-powers commented on src/main/java/frc/robot/subsystems/ArmSubsystem.java:
```diff
@@ -0,0 +1,49 @@
+    public void periodic() {
+    }
+
+    public void adjustShoulders(double power) {
+        power *= MoPrefs.shoulderSetpointRpm.get();
+
+        leftShoulderVelocityPID.setReference(power);
+        rightShoulderVelocityPID.setReference(power);
+    }
```
The whole point of a fallback "Direct" mode is for when we don't have functional, tuned PID.

You should be direct-driving the motors here, using [CANSparkMax.set()](https://codedocs.revrobotics.com/java/com/revrobotics/cansparkmax#set(double)).

The same applies to `adjustWrist()`

dslit3 committed:
Revisions and functionality changes

- Allow picking between PID and non-PID for arm control
- Arm mode can be chosen during robot run (not just reboot)
- Shoulder motor controllers follow each other
- Add deadzone and curve to arm inputs
- Code cleanup

dslit3 committed:
Fix build

jordan-powers commented on src/main/java/frc/robot/subsystems/ArmSubsystem.java:
```diff
@@ -0,0 +1,89 @@
+     * @param target -1.0 to 1.0, each being the max speed in that direction
+     */
+    public void adjustShoulders(boolean pid, double target) {
+        target *= MoPrefs.shoulderSetpointRpm.get();
```
This looks like a bug: target should only be converted to velocity when in PID mode.
`CANSparkMax.set()` takes a direct motor power between [-1,1]. We don't want to convert to velocity (which has the range [-MAX_VELOCITY, MAX_VELOCITY] if we're trying to set motor power.

The same comment applies to `adjustWrist()`

jordan-powers reviewed (commented):
Looks good! Just fix the one bug I noticed, and we're good to merge!

jordan-powers commented on src/main/java/frc/robot/subsystems/ArmSubsystem.java:
```diff
@@ -0,0 +1,89 @@
+    private final TeleopArmCommand pidArmControlCommand;
+
+    public ArmSubsystem(MoInput input) {
+        rightShoulder.setInverted(true); // TODO: Verify
+        rightShoulder.follow(leftShoulder);
```
One nitpick here: there's an overload of `.follow()` that allows you to specify that it should follow inverted. Might be a bit cleaner than two lines, but (assuming `.follow()` respects a previous call to `.setInverted()`) it's not a big deal.

momentumrobotics-4999 committed:
Fix non-PID arm movement and update WPILib
